### ISSUE TITLE
Let rebar3 plugin check beam instead of erl files

### ISSUE THREAD

**Issue by zuiderkwast** [josefs/Gradualizer#300]
I tried the rebar3 plugin on a project today. I ran into a few problems.

1. I get a lot of error like "Undefined type options/0 on line 0", which I assume is because the "include/" directory is not included when parsing the erl files (would be solved by checking beams instead, but it would also be possible to do better guessing of include dirs). (Maybe I could solve this by setting `gradualizer_opts` in rebar.config, but working out of the box would be nicer.)
2. The output doesn't contain the names of the files being checked, making it hard to see what each error is referring to.
3. I would like to make Gradualizer a dependency in my lib only when running checks, not when only compiling and when using my lib as a dependency in another project. Is this possible already?

**Comment by ilya-klyuchnikov** [comment]
FYI: I am doing some hacking of gradualizer on my own (for some practical stuff) - without any intention to upstream it so far, but - it was the first thing I did - https://github.com/ilya-klyuchnikov/Gradualizer/commit/560cac7d5b49d1fb9c0efc3592028288eaba22de - feel free to port it/upstream it on your own)

**Comment by zuiderkwast** [comment]
Nice @ilya-klyuchnikov! Is it fully working or is there something more needed? Why not create PRs? It looks like your doing good stuff.

@Zalastax, do you think we can merge this commit as-is?

**Comment by Zalastax** [comment]
Yes, I think the commit looks good! I don't know why the more complicated way was chosen.
________________________________

**Comment by ilya-klyuchnikov** [comment]
I am doing some things with it for Whatsapp in the first place (and it works for WA in some pilot/research mode) - let me have a chat with @josefs of how to bring it upstream effectively if there is a community interest in it.

**Comment by zuiderkwast** [comment]
Sounds good. So you know @josefs? I could have guessed. :)

**Comment by ilya-klyuchnikov** [comment]
OK, - I had a quick chat with @josefs , - it would be beneficial to upstream changes/improvements. - I will make corresponding PRs this/next week. Thanks everyone for the discussion.

**Comment by zuiderkwast** [comment]
The PR adds an option to gradualize from beams in the rebar3 plugin :tada:, but there are still issues after merging that PR:
* Beams are not default
* When using erls, include files are not found. I found out rebar3 has some hard-coded defaults for finding include files when compiling. If we are to support erls, we should use the same logic as rebar3's compile target.
    ```Erlang
    src/rebar_erlc_compiler.erl-internal_erl_compile(Opts, Dir, Module, OutDir, ErlOpts, RebarOpts) ->
    src/rebar_erlc_compiler.erl-    Target = target_base(OutDir, Module) ++ ".beam",
    src/rebar_erlc_compiler.erl-    ok = filelib:ensure_dir(Target),
    src/rebar_erlc_compiler.erl-    PrivIncludes = [{i, filename:join(Dir, Src)}
    src/rebar_erlc_compiler.erl-                    || Src <- rebar_dir:all_src_dirs(RebarOpts, ["src"], [])],
    src/rebar_erlc_compiler.erl-    AllOpts = [{outdir, filename:dirname(Target)}] ++ ErlOpts ++ PrivIncludes ++
    src/rebar_erlc_compiler.erl-              [{i, filename:join(Dir, "include")}, {i, Dir}, return],
    src/rebar_erlc_compiler.erl:    case compile:file(Module, AllOpts) of
    ```
* Filenames should be displayed by default, at least when checking more than one file, as this is the default behaviour for the standalone escript. (Off topic: Beam filenames like "my_mod.beam" doesn't look that nice since the line numbers refer to the source file. I think we should display the filename stored in the AST instead, not only for the rebar3 plugin but always.)

**Comment by ilya-klyuchnikov** [comment]
> Filenames should be displayed by default

I think that they ALWAYS should be displayed in the format
```
filename:Line
filename:Line:Column
```

Why?
- This is what the majority of the compilers/checkers do
- The majority of editors/terminals/IDEs understand this format - making these `filename:Line` clickable - so you can easily jump to the error location (with line)

I would argue to have the only mode with displaying filenames. - This is what users expect. (And this helps keep gradualizer code simple - fewer options/code paths - in the first place).

The only case when one may want to NOT have a filename - is when developing/hacking on gradualizer itself. I am not sure whether it's such a distraction of having filenames in console.

**Comment by zuiderkwast** [comment]
I would accept a PR deleting all the other options.

We have all these options just be because we didn't want to take decisions. :-) But it's hard to maintain the combinations.
